Mathlib/Analysis/ODE/PicardLindelof.lean

```diff
@@ -10,397 +10,716 @@ import Mathlib.Topology.MetricSpace.Contracting
 
 ## Tags
 
-differential equation
+differential equation, dynamical system, initial value problem, Picard-Lindelöf theorem,
+Cauchy-Lipschitz theorem
+
 -/
 
-open Filter Function Set Metric TopologicalSpace intervalIntegral MeasureTheory
-open MeasureTheory.MeasureSpace (volume)
-open scoped Filter Topology NNReal ENNReal Nat Interval
+open Function intervalIntegral MeasureTheory Metric Set
+open scoped Nat NNReal Topology
+
+/-! ## Assumptions of the Picard-Lindelof theorem-/
+
+/-- Prop structure holding the assumptions of the Picard-Lindelöf theorem.
+`IsPicardLindelof f t₀ x₀ a r L K`, where `t₀ ∈ Icc tmin tmax`, means that the time-dependent vector
+field `f` satisfies the conditions to admit an integral curve `α : ℝ → E` to `f` defined on
+`Icc tmin tmax` with the initial condition `α t₀ = x`, where `‖x - x₀‖ ≤ r`. Note that the initial
+point `x` is allowed to differ from the point `x₀` about which the conditions on `f` are stated. -/
+structure IsPicardLindelof {E : Type*} [NormedAddCommGroup E]
+    (f : ℝ → E → E) {tmin tmax : ℝ} (t₀ : Icc tmin tmax) (x₀ : E) (a r L K : ℝ≥0) : Prop where
+  /-- The vector field at any time is Lipschitz with constant `K` within a closed ball. -/
+  lipschitzOnWith : ∀ t ∈ Icc tmin tmax, LipschitzOnWith K (f t) (closedBall x₀ a)
+  /-- The vector field is continuous in time within a closed ball. -/
+  continuousOn : ∀ x ∈ closedBall x₀ a, ContinuousOn (f · x) (Icc tmin tmax)
+  /-- `L` is an upper bound of the norm of the vector field. -/
+  norm_le : ∀ t ∈ Icc tmin tmax, ∀ x ∈ closedBall x₀ a, ‖f t x‖ ≤ L
+  /-- The time interval of validity -/
+  mul_max_le : L * max (tmax - t₀) (t₀ - tmin) ≤ a - r
+
+namespace ODE
+
+/-! ## Integral equation
+
+For any time-dependent vector field `f : ℝ → E → E`, we define an integral equation that is
+equivalent to the initial value problem defined by `f`.
+-/
 
-noncomputable section
+section
 
 variable {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E]
+  {f : ℝ → E → E} {α : ℝ → E} {s : Set ℝ} {u : Set E} {t₀ tmin tmax : ℝ}
+
+/-- The Picard iteration. It will be shown that if `α : ℝ → E` and `picard f t₀ x₀ α` agree on an
+interval containing `t₀`, then `α` is a solution to `f` with `α t₀ = x₀` on this interval. -/
+noncomputable def picard (f : ℝ → E → E) (t₀ : ℝ) (x₀ : E) (α : ℝ → E) : ℝ → E :=
+  fun t ↦ x₀ + ∫ τ in t₀..t, f τ (α τ)
+
+@[simp]
+lemma picard_apply {x₀ : E} {t : ℝ} : picard f t₀ x₀ α t = x₀ + ∫ τ in t₀..t, f τ (α τ) := rfl
+
+lemma picard_apply₀ {x₀ : E} : picard f t₀ x₀ α t₀ = x₀ := by simp
+
+/-- Given a $C^n$ time-dependent vector field `f` and a $C^n$ curve `α`, the composition `f t (α t)`
+is $C^n$ in `t`. -/
+lemma contDiffOn_comp {n : WithTop ℕ∞}
+    (hf : ContDiffOn ℝ n (uncurry f) (s ×ˢ u))
+    (hα : ContDiffOn ℝ n α s) (hmem : ∀ t ∈ s, α t ∈ u) :
+    ContDiffOn ℝ n (fun t ↦ f t (α t)) s := by
+  have : (fun t ↦ f t (α t)) = (uncurry f) ∘ fun t ↦ (t, α t) := rfl -- should this be a lemma?
+  rw [this]
```

Comment:
In several projects, I have been ask to take a stand and either implement or remove this kind of comments (which, actually, made sense to me…). If you use it more than once, I'dd suggest to write a lemma. I also find the following syntax useful.
```suggestion
  rw [show fun t ↦ f t (α t)) = (uncurry f) ∘ fun t ↦ (t, α t) by rfl]
```